### DISCUSSION TITLE
Callback has no derivative

### DISCUSSION THREAD

**Discussion by mrrezaie** [casadi/casadi#3580] [Q&A] [answered]
Hi, I have a Callback function with the following form and want to defined an equality constraint with its outputs:
```python
class myFunction(casadi.Callback):
    def __init__(self, name, opts={}):
        casadi.Callback.__init__(self)
        self.construct(name, opts)

    # Number of inputs and outputs
    def get_n_in(self):  return 6
    def get_n_out(self): return 2

    # Initialize the object
    def init(self): print('Initializing the object')

    # Evaluate the function numerically
    def eval(self, arg):

        act  = float(arg[0])
        NML  = float(arg[1])
        NMV  = float(arg[2])
        MLAT = float(arg[3])
        MTL  = float(arg[4])
        mi   = int(  arg[5])

        # some numerical processing with non-casadi functions
        # TF and MFAT are the output (float)

        return TF, MFAT

f = myFunction('f')
```
But solver raises this error:

`RuntimeError: Error in Opti::solve [OptiNode] at .../casadi/core/optistack.cpp:157:
Error in MX::jtimes at .../casadi/core/generic_matrix.hpp:1456:
Error in MX::forward at .../casadi/core/mx.cpp:1755:
Error in MXFunction::ad_forward at .../casadi/core/mx_function.cpp:890:
Error in MX::ad_forward for node of type N6casadi4CallE at .../casadi/core/mx.cpp:2301:
Error in Call::ad_forward for 'f' [CallbackInternal] at .../casadi/core/casadi_call.cpp:121:
.../casadi/core/function_internal.cpp:2879: Assertion "has_derivative()" failed:
Derivatives cannot be calculated for f`

I saw the [callback examples](https://github.com/casadi/casadi/blob/main/docs/examples/python/callback.py) and found that I should add extra functions to the Callback such as jacobian and sparsity, but I don't know how to proceed. `{"enable_fd":True}` also got it working but never solved/converged. Any idea how to work with this Callback and fix this issue?

Thank you in advance.

**Comment by brunoml5** [comment, marked as answer]
Hi!

The main reason for using tools such as Casadi is to write models with efficient derivative information since it is used a lot in optimization. When you write a Callback function without some kind of derivative, Casadi fails as you saw. You can enable finite differences as you tried but this is a terrible way of calculating the derivative. It is ineficient and inacurrate.

So, when building your function, it is important to supply some kind derivative in forward, reverse or full jacobian. Take a look at this [blog post](https://web.casadi.org/blog/tensorflow/) using Tensorflow with Casadi, it is quite interesting. 

To learn more about algorithmic differentiation, take a look at this [video](https://www.youtube.com/watch?v=mYOkLkS5yqc) from Joris Gillis which is one of the main developers of Casadi. There are other helpful videos in his channel as well.

I hope it helps!

PS: I'm not a developer.

**Reply by mrrezaie** [reply]
Hi, thanks for your response.

Yes, I've been trying to understand how to write full jacobian for this callback, but it's actually so complicated for me.
